### DISCUSSION TITLE
Working with pyinstaller and singnals (SIGTSTP)

### DISCUSSION THREAD

**Discussion by gammore** [pyinstaller/pyinstaller#8341] [Help] [answered]
---
Versions:
pyinstaller==6.4.0
pyinstaller-hooks-contrib==2024.2

python 3.11.7

---

Hi! 

I'm encountering a weird issue with pyisntaller and my script. I have a python script that executes a subprocess in a interactive way. I want the process to handle the signals (CTRL+Z, CTRL+C, etc) but instead when I press those keys the python execution hangs.

My program does something like this (simplified):
```python
(...)
cmd = f"ssh -i {key} {user}@{server}"
signal.signal(signal.SIGINT, signal.SIG_IGN)
signal.signal(signal.SIGQUIT, signal.SIG_IGN)
signal.signal(signal.SIGTSTP, signal.SIG_IGN)
subprocess.run(cmd, shell=True)
(...)
```

The idea is that python runs the command and then all signals (CTRL+Z, CTRL+C, etc) are handled by the command (ssh). This way, if I go into a server using this, i can CTRL+C a command i'm running:
```
python mycommand.py -k key -u user -s server
(...)
(ssh) server@user-$ # ssh commands here
```

What I'm seeing is exactly the same as this comment on stack overflow:
https://stackoverflow.com/questions/76550274/executable-generated-from-python-script-with-pyinstaller-is-not-handling-ctrlz

This doesn't happen if I run the command with python:
```bash
python mycommand.py -k key -u user -s server   # works as expected with CTRL+Z
./dist/mycommand -k key -u user -s server           # exits python whith CTRL+Z
```

Do you have any ideas as to why this happens?

**Comment by gammore** [comment]
Note that CTRL + C is handled correctly, while logged in i can press ctrl c and the process doesn't end:
```
[root@xxx ~]#
[root@xxx ~]# ^C
[root@xxx ~]# ^C
```

Also I've tried using the flag `--bootloader-ignore-signals`, but still doesn't work:
```
[root@xxx ~]#  # (pressed ctrl-z here)
[1]  + 1137104 suspended  ./dist/mycommand -k key -u user -s server
```

This is the command i use to buld the package:
```
pyinstaller --onefile --bootloader-ignore-signals mycommand.py
```

**Comment by gammore** [comment]
I've seen this, but doesn't seem fixed:
https://github.com/pyinstaller/pyinstaller/pull/4244

**Comment by rokm** [comment, marked as answer]
Signals are somewhat messy in `onefile` builds, because it has a parent and a child process, and the semantics of passing the signals between the two are ill-posed in some cases, due to issue of sending signal to parent process (e.g., via `kill -SIGNUM <pid>`) and sending signal to whole process group (which is what pressing Ctrl+C or Ctrl+Z in your terminal actually does).

I think `onedir` build of your example program would behave as you expect, i.e., similar to unfrozen python. Your python code disables the signal handling in program process, so if you send SIGINT or SIGTSTP to it via `kill`, nothing happens. That's also why if you press Ctrl+C or Ctrl+Z, the program ignores the signal it received via process group, while the subprocess running shell receives it, and acts on it.

In a `onefile` build, we have a parent process that unpacks itself to temporary directory, then runs a child instance that uses the unpacked files. In this mode, most signals sent to the parent should be ignored and forwarded to the child instead. For example, if you send SIGINT via `kill` to the parent process, the parent should send signal to the child, and if the child exits as a result (i.e., it is not blocking the signal itself), clean the temporary files (i.e., the parent should keep running in spite of SIGINT).

On the other hand, if you press Ctrl+C, both parent and child receive the signal (because the signal was sent to process group, and both those processes are in it). But parent ignores the signal and forwards it to the child, who now receives two signals - one from process group and one from the parent. This can, depending on timing, lead to exception being thrown while python attempts to handle the signal. And I think the `--bootloader-ignore-signals` flag was added to work around this problem - it installs different signal handlers that do nothing, and thus prevents parent process from forwarding signals to the child. So in the case of Ctrl+C, the child always receives only one signal. On the flip side, if you send SIGINT directly to the parent process via `kill`, nothing will happen due to lack of forwarding.

Which brings us to Ctrl+Z and SIGTSTP. If you wanted to stop a onefile application, you would need to stop both parent and child process. Which means that the PyInstaller's default behavior of ignoring the signal in parent and forwarding it to the child is wrong - that's why  #4244 added SIGSTP to the list of exceptions for which signal handlers in the parent process are not modified. This ensures that if Ctrl+Z is pressed, both processes are suspended. It does not work if you send SIGTSTP directly to parent, though; you will suspend the parent, but not the child. On the flip side, though, it prevents the child process from receiving two SIGTSTP signals when Ctrl+Z is pressed (which might or might not be problematic in a similar way as it is from Ctrl+C).

And lastly, your use-case is different from what was discussed so far. You want SIGTSTP handling disabled in both the parent process and in the child process, so that when SIGTSTP is sent to process group on Ctrl+Z, none of the processes are suspended, and only the spawned ssh program acts upon it. And while you have successfully disabled signal handling in the child process (via python's `signal.signal` calls), you currently have no means of doing so for the parent process. `--bootloader-ignore-signals` *should* do that for you, but it does not, because #4244 exemtped SIGTSTP from having custom signal handler installed at all (i.e., both when `--bootloader-ignore-signals` is specified, and when it is not).

So if you really want to achieve the behavior you describe, and want to use `onefile` mode, you should probably modify this part of bootloader:

https://github.com/pyinstaller/pyinstaller/blob/18d5325e871865e24f194a04ebc88968ad84a0ee/bootloader/src/pyi_utils.c#L1070-L1077

and ensure that the custom [`_ignoring_signal_handler`](https://github.com/pyinstaller/pyinstaller/blob/18d5325e871865e24f194a04ebc88968ad84a0ee/bootloader/src/pyi_utils.c#L979)
signal handler is installed for SIGTSTP when `ignore_signals` is set.

**Comment by gammore** [comment]
Hi!

Thank you very much for your answer, now I understand it a lot more! 

I think I was getting obsessed with having the isntallation in one file only, but it's not that practical anyway. I'm going to try with your approach and test!
